### PR TITLE
[WIP] Standard deployment for cloujera

### PR DESCRIPTION
**Pull request by vise890**
This PR aims at streamlining the deployment to prod as well as the local dev setup.
- Vagrant can be used locally for running elasticsearch and redis (both running in Docker containers inside the VM)
  - the ports are forwarded so you can use this VM instead of having to configure and install elasticsearch and redis locally
- `provision.sh` can be used to provision a production machine running ubuntu for the first time
- `deploy.sh` provides easy deployment on said machine

This allows deployment to the cloud to be pretty much: 

``` shell
$ ssh user@cloudmachine.com
$ git checkout https://github.com/clojurecup2014/cloujera
$ cd cloujera
$ ./scripts/provision.sh
$ ./scripts/deploy.sh
```
## LEFT TODO:
- Cloujera still can't connect to redis from inside its container (investigating iptables issue)
## BUGS

There is one very sore note. `burglar.core` still makes an HTTP call at **compile time**. But I haven't been able to eliminate that while keeping coursera happy (got throttled, banned, etc.). This is bad because sometime compilation fails because of the network........

### CODE REVIEW

File: resources/public/index.html [hunk 1]
```diff
@@ -71,8 +71,17 @@
     </div>
 
     <script src="//cdnjs.cloudflare.com/ajax/libs/react/0.9.0/react.min.js"></script>
+
+    <!-- DEV version -->
+    <script src="js/out-dev/goog/base.js" type="text/javascript"></script>
+    <script src="js/cloujera-dev.js" type="text/javascript"></script>
+    <!-- end DEV version -->
+
+    <!-- PROD version
```
Comment (vise890) [inline resources/public/index.html]: this still needs to be commented in by hand to deploy to prod. I'm talking with one of the guys at hacker school tomorrw, he knows Om/cljs and should know what to do.

File: project.clj [hunk 2]
```diff
@@ -33,14 +33,18 @@
...
-  :cljsbuild {
-    :builds [{:id "cloujera"
-              :source-paths ["src/cljs"]
-              :compiler {
-                :output-to "resources/public/js/cloujera.js"
-                :output-dir "resources/public/js/out"
-                :optimizations :none
-                :source-map true}}]}
+  :cljsbuild {:builds [{:id "dev"
+                        :source-paths ["src/cljs" "dev-config"]
```
Comment (vise890) [inline project.clj]: the config for the `cljs` backend URI now lives in files in `./dev-config` and `./prod-config`

**Comment by kgxsz** [comment]
Me gusta. Why do you not want to merge it?

**Comment by vise890** [comment]
Its still broken :(

​I'll work on it today too tho.​

**Comment by vise890** [comment]
Ok guys i think this is almost ready for merge. The very last thing. 

I am building a container for cloujera and running it linked to both the `redis` and `elasticsearch` containers. But it doesn't seem like the `cloujera` container can talk to them.

If I try `docker exec cloujera curl localhost:9200` I get a `Connection refused` error. (NB: 9200 is a port exposed by elasticsearch, that gives a JSON response over HTTP)

On the other hand, running `curl localhost:9200` in the VM or in the guest, I get the JSON. So it's definitely and inter-container communication problem.

@kgxsz can you have a look at this? I am linking the containers in the `deploy.sh` script..

**Comment by vise890** [comment]
Ok I think I have found the cause and a way to fix it. I'll do it tomorrow if i can. I realise this is probably overengineered, but it's teaching me tons...

**Comment by kgxsz** [comment]
@vise890 is it possible that the `docker exec` command runs the command from the perspective of inside the container, such that the localhost is not referring to the VM itself, but rather the container. From memory, and this is a longshot, you may have to do some port forwarding magic when you start the container.

**Comment by vise890** [comment]
So, from inside the container, the cloujera app was trying to connect on elasticsearch/redis on `localhost`.

But docker assigns different IPs to the elasticsearch and redis contaners, so cloujera should connect to those.

you can see a WIP fix (I haven't tested it yet) in c33088f where the `deploy.sh` asks docker for the container IPs and exports them. Then I use `environ` to look the values up in the app.

**Comment by kgxsz** [comment]
@vise890 that sounds right - alternatively, you may want to look into port forwarding from host to container, it could solve the problem without having to issue env vars.

**Comment by vise890** [comment]
i think it's cool, if you link containers A, B to container C say, you
actually get an environment variable <C>_PORT=tcp://<C'S_IP>:<C'S_PORT>
which I am now using. I broke the cache now tho. But it should be almost
ready.

**Comment by vise890** [comment]
Guys, I am completely stuck. I have tried everything I could think of. I can't get the app to work from inside a Docker container.

Cloujera connects to `elasticsearch`, but not to `redis`. I have no idea why. I just get a connection refused error every time. If you guys wanna download this thing and take it for a spin, feel free. The `README.md` should be pretty comprehensive. But I'm closing the PR. I think it's an overengineered dead end.

Things i've tried (and didn't work):
- disabling firewall `ufw disable`
- running Docker with full inter-container communication `--icc=true`
- running bash inside a the `cloujera` container and trying to connect to redis from `lein repl`

Things that do work:
- running the `uberjar` straight in the VM after exporting the env vars manually (as per the readme)
- connecting to `redis`, while inside the cloujera container with `redis-cli -h XX` (where XX is the `redis` container IP)
- `lein repl` straight in the Ubuntu VM (not in the Container), vith env vars exported

The redis configuration gets picked up correctly in the container from the environment variables. It just can't connect.

Here's a diagram of the infrastructure, more for posterity than anything else:
infra_diagram

**Comment by kgxsz** [comment]
I'll have a look this week.

**Comment by OnesAndTwos** [comment]
I'll have a look as well :)

**Comment by vise890** [comment]
Ok so I'm working on it today with a friend who's really interested in it.
I'm merging the small cleanups and things that came out of the spike. We're
gonna try again from scratch.

**Pull request closed without merging**